### PR TITLE
Add formatting for `StmtMatch`

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR adds support for `StmtMatch` with subs for `MatchCase`.

## Test Plan

Add a few additional test cases around `match` statement, comments, line breaks.

resolves: #6298

### CODE REVIEW

**Comment by dhruvmanila** [comment]
Current dependencies on/for this PR:
* main
  * **PR #6286** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/6286" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/6286?utm_source=stack-comment).

File: crates/ruff_python_formatter/tests/snapshots/format@statement__match.py.snap [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+```py
+# leading match comment
+match foo:  # dangling match comment
+    case NOT_YET_IMPLEMENTED_MatchCase:
+        pass
+
+
+# leading match comment
+match (
+    # leading expr comment
```
Comment (dhruvmanila) [inline crates/ruff_python_formatter/tests/snapshots/format@statement__match.py.snap:76]: This seems to be a known issue where the comment right next to an opening parentheses is moved down instead.

```diff
- match (  # leading expr comment
+ match (
+     # leading expr comment
```

File: crates/ruff_python_formatter/src/comments/placement.rs [hunk 2]
```diff
@@ -419,7 +425,7 @@ fn handle_own_line_comment_in_clause<'a>(
 ///     # Leading `case (x, y, z)` comment
 ///     case _:
 /// ```
-fn handle_match_comment<'a>(
+fn handle_match_case_comment<'a>(
```
Comment (konstin) [inline crates/ruff_python_formatter/src/comments/placement.rs:428]: Do we need this function? i left it because i hadn't worked with match yet, but that case should be handled by the generic if/while/for/try logic

File: crates/ruff_python_formatter/src/statement/stmt_match.rs [hunk 3]
```diff
@@ -1,12 +1,55 @@
...
+        let dangling_item_comments = comments.dangling_comments(item);
+
+        write!(
+            f,
+            [
+                text("match"),
+                space(),
+                maybe_parenthesize_expression(subject, item, Parenthesize::IfBreaks),
+                text(":"),
+                dangling_comments(dangling_item_comments)
```
Comment (konstin) [inline crates/ruff_python_formatter/src/statement/stmt_match.rs:32]: Can this be more than one comment, and if not, could you add a debug assert?

**Review by konstin — APPROVED** [review]

File: crates/ruff_python_formatter/src/statement/stmt_match.rs [hunk 4]
```diff
@@ -1,12 +1,55 @@
...
+            write!(
+                f,
+                [block_indent(&format_args![
+                    text("case"),
+                    space(),
+                    not_yet_implemented_custom_text("NOT_YET_IMPLEMENTED_MatchCase"),
+                    text(":"),
+                    block_indent(&case.body.format())
+                ])]
+            )?;
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/stmt_match.rs:46]: Could we move the mock implementations to the match cases instead and call into `case.format().fmt(f)` here? 

This has the benefit that implementing a specific match case requires no changes to the statement formatting (and we can even provide match case mocks of the right pattern)

File: crates/ruff_python_formatter/src/statement/stmt_match.rs [hunk 5]
```diff
@@ -1,12 +1,55 @@
...
+                text("match"),
+                space(),
+                maybe_parenthesize_expression(subject, item, Parenthesize::IfBreaks),
+                text(":"),
+                dangling_comments(dangling_item_comments)
+            ]
+        )?;
+
+        for case in cases {
+            write!(
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/stmt_match.rs:37]: Do we need to retain empty lines between match cases?

Micha trying to write valid python code :sweat_smile: 

```python
match test:
	case "a":
		pass

	case "b":
		pass
```

File: crates/ruff_python_formatter/tests/snapshots/format@statement__match.py.snap (hunk 1 above)
Comment (konstin) [reply crates/ruff_python_formatter/tests/snapshots/format@statement__match.py.snap:76]: See also #6380 and #6381, but we don't have the handling for this for general expression parentheses yet, leading end-of-line comments inside of optional expression parentheses are rather rare

File: crates/ruff_python_formatter/src/statement/stmt_match.rs (hunk 4 above)
Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/statement/stmt_match.rs:46]: Yeah, we can.

> (and we can even provide match case mocks of the right pattern)

By this, do you mean the different kinds of patterns for the match cases? If so, that will make the formatting unstable because in the first iteration each pattern type will be formatted as it's not implemented equivalent but on the second iteration (now that the pattern is a variable semantically), some of them will be formatted again.

```python
match p:
	case Point(x, y):  # Pattern::MatchClass
		pass

# vvvvv
match p:
	case NOT_YET_IMPLEMENTED_MatchClass:  # Pattern::MatchValue
		pass

# vvvvv
match p:
	case NOT_YET_IMPLEMENTED_MatchValue:
		pass
```

File: crates/ruff_python_formatter/src/statement/stmt_match.rs (hunk 5 above)
Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/statement/stmt_match.rs:37]: I think the newline handling will be done in the `MatchCase` implementation. Any newline between the `match` statement and the first case will be removed as is done in black as well:

```python
match test:

	case "a":
		pass

# vvvvv
match test:
	case "a":
		pass
```

File: crates/ruff_python_formatter/src/statement/stmt_match.rs (hunk 4 above)
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/statement/stmt_match.rs]: You can use the not_implemented_custom_text to provide a specific text. You then need to make sure that the class item uses a class verbatim text.

If it's not already the case: The MatchCase formatting should be a match statement that dispatches the formatting to the specific case item (see FormatStmt)

File: crates/ruff_python_formatter/src/comments/visitor.rs [hunk 6]
```diff
@@ -73,7 +73,6 @@ impl<'a> CommentsVisitor<'a> {
                 enclosing: enclosing_node,
                 preceding: self.preceding_node,
                 following: Some(node),
-                parent: self.parents.iter().rev().nth(1).copied(),
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/comments/visitor.rs:76]: Nice!

File: crates/ruff_python_formatter/src/other/match_case.rs [hunk 7]
```diff
@@ -1,12 +1,45 @@
...
+            guard,
+            body,
+        } = item;
+
+        write!(
+            f,
+            [
+                text("case"),
+                space(),
+                not_yet_implemented_custom_text("NOT_YET_IMPLEMENTED_Pattern"),
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/other/match_case.rs:27]: We can also do this as a separate PR: Ideally, we call into `pattern.format()` here. This requires implementing `AsFormat` and `IntoFormat` for `Pattern`, similar to how it is done for `Mod`

https://github.com/astral-sh/ruff/blob/f45e8645d7c4a2ad4e934ebbbadd91ca59fd543a/crates/ruff_python_formatter/src/module/mod.rs#L1-L35

**Review by MichaReiser — APPROVED** [review]

File: crates/ruff_python_formatter/src/other/match_case.rs (hunk 7 above)
Comment (dhruvmanila) [reply crates/ruff_python_formatter/src/other/match_case.rs:27]: Yes, it's already a work in progress in the next PR (https://github.com/astral-sh/ruff/pull/6360/files#diff-05760c9829e7694c86aaf97014143c0785c12da546f06dcfbc87caf67b15d773) ;)
